interface not being interpreted

### The following program `sample.go` triggers an unexpected result

```go
package xpath

import (
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
	"sync"
	"unicode"
)

// Defined an interface of stringBuilder that compatible with
// strings.Builder(go 1.10) and bytes.Buffer(< go 1.10)
type stringBuilder interface {
	WriteRune(r rune) (n int, err error)
	WriteString(s string) (int, error)
	Reset()
	Grow(n int)
	String() string
}

var builderPool = sync.Pool{New: func() interface{} {
	return newStringBuilder()
}}
//...
```


### Expected result

```console
it should at least compile
```


### Got

```console
jptosso@Porco-Rosso  ~/go/coraza-traefik   master  /Users/jptosso/go/bin/yaegi test  
/Users/jptosso/go/src/github.com/jptosso/coraza-waf/vendor/github.com/antchfx/xpath/func.go:23:19: panic
panic: reflect.Set: value of type func(interface {}, string, interface {}, interface {}) bool is not assignable to type *interp.node [recovered]
	panic: reflect.Set: value of type func(interface {}, string, interface {}, interface {}) bool is not assignable to type *interp.node

goroutine 1 [running]:
github.com/traefik/yaegi/interp.runCfg.func1(0xc0003e4000, 0xc0017ab100, 0xc000b92e38)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/run.go:184 +0x253
panic(0x1a1e180, 0xc00179a5b0)
	/usr/local/Cellar/go/1.16.5/libexec/src/runtime/panic.go:965 +0x1b9
reflect.Value.assignTo(0xc0016b2b00, 0xc0005af670, 0x193, 0x1b5f588, 0xb, 0x1af92c0, 0x0, 0xc0016b2b00, 0xc0005af638, 0xc0016b2b00)
	/usr/local/Cellar/go/1.16.5/libexec/src/reflect/value.go:2451 +0x3f7
reflect.Value.Set(0x1af92c0, 0xc000a74e68, 0x196, 0xc0016b2b00, 0xc0005af670, 0x193)
	/usr/local/Cellar/go/1.16.5/libexec/src/reflect/value.go:1564 +0xbd
github.com/traefik/yaegi/interp.arrayLit.func1(0xc0003e4000, 0xc00213a930)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/run.go:2338 +0x16e
github.com/traefik/yaegi/interp.runCfg(0xc0017ab100, 0xc0003e4000)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/run.go:190 +0x87
github.com/traefik/yaegi/interp.(*Interpreter).run(0xc0003e2000, 0xc001202500, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/run.go:121 +0x2b0
github.com/traefik/yaegi/interp.(*Interpreter).importSrc(0xc0003e2000, 0xc000f2d710, 0x24, 0xc00168be21, 0x18, 0x1, 0xc001234b10, 0x3, 0x0, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/src.go:148 +0x9d4
github.com/traefik/yaegi/interp.(*Interpreter).gta.func1(0xc001408e00, 0xc001225c28)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/gta.go:226 +0x13d0
github.com/traefik/yaegi/interp.(*node).Walk(0xc001408e00, 0xc000b93c28, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:233 +0xb5
github.com/traefik/yaegi/interp.(*node).Walk(0xc001408900, 0xc000b93c28, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:237 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc001408700, 0xc000b93c28, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:237 +0x66
github.com/traefik/yaegi/interp.(*Interpreter).gta(0xc0003e2000, 0xc001408700, 0xc001189b80, 0x40, 0xc00168b7c1, 0x1b, 0x0, 0x1, 0xc001408700, 0x0, ...)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/gta.go:21 +0x21f
github.com/traefik/yaegi/interp.(*Interpreter).importSrc(0xc0003e2000, 0xc000b15530, 0x24, 0xc00168b7c1, 0x1b, 0x1, 0xc0013fea10, 0xc0012263a8, 0x13986c6, 0xc000263610)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/src.go:100 +0xf57
github.com/traefik/yaegi/interp.(*Interpreter).gta.func1(0xc001238500, 0xc0012269a8)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/gta.go:226 +0x13d0
github.com/traefik/yaegi/interp.(*node).Walk(0xc001238500, 0xc000b949a8, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:233 +0xb5
github.com/traefik/yaegi/interp.(*node).Walk(0xc001202c00, 0xc000b949a8, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:237 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc001202a00, 0xc000b949a8, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:237 +0x66
github.com/traefik/yaegi/interp.(*Interpreter).gta(0xc0003e2000, 0xc001202a00, 0xc00168bdc0, 0x1d, 0xc000508081, 0x1d, 0x4, 0x8, 0xc001202a00, 0x0, ...)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/gta.go:21 +0x21f
github.com/traefik/yaegi/interp.(*Interpreter).importSrc(0xc0003e2000, 0xc000291dba, 0x1, 0xc000508081, 0x1d, 0xc0003e2001, 0xc0002848f0, 0xc0002a27c0, 0xd4, 0x119b333)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/src.go:100 +0xf57
github.com/traefik/yaegi/interp.(*Interpreter).gta.func1(0xc00050ad00, 0xc0001a9728)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/gta.go:226 +0x13d0
github.com/traefik/yaegi/interp.(*node).Walk(0xc00050ad00, 0xc000b95728, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:233 +0xb5
github.com/traefik/yaegi/interp.(*node).Walk(0xc00050a400, 0xc000b95728, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:237 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc00050a200, 0xc000b95728, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:237 +0x66
github.com/traefik/yaegi/interp.(*Interpreter).gta(0xc0003e2000, 0xc00050a200, 0xc000291dba, 0x1, 0xc00029196c, 0x2, 0x0, 0x1, 0xc00050a200, 0x0, ...)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/gta.go:21 +0x21f
github.com/traefik/yaegi/interp.(*Interpreter).importSrc(0xc0003e2000, 0x1cd6cd8, 0x1, 0xc00029196c, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/src.go:100 +0xf57
github.com/traefik/yaegi/interp.(*Interpreter).EvalTest(...)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/interp/interp.go:427
main.test(0xc000032060, 0x0, 0x0, 0xc00000a601, 0xc000001680)
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/cmd/yaegi/test.go:151 +0x106a
main.main()
	/Users/jptosso/go/pkg/mod/github.com/traefik/yaegi@v0.9.23/cmd/yaegi/yaegi.go:135 +0x505
```

It's some issue with sync.Pool{} which takes New(...) interface{} from repo github.com/antchfx/xpath (func.go)


### Yaegi Version

devel (latest)

### Additional Notes

I see **Interfaces to be used from the pre-compiled code can not be added dynamically, as it is required to pre-compile interface wrappers.** in the readme, is there any workaround or any way I can fix this?

Thank you

--- a/_test/assert2.go
+++ b/_test/assert2.go
@@ -1,0 +1,37 @@
+package main
+
+import (
+	"strings"
+	"sync"
+)
+
+// Defined an interface of stringBuilder that compatible with
+// strings.Builder(go 1.10) and bytes.Buffer(< go 1.10)
+type stringBuilder interface {
+	WriteRune(r rune) (n int, err error)
+	WriteString(s string) (int, error)
+	Reset()
+	Grow(n int)
+	String() string
+}
+
+var builderPool = sync.Pool{New: func() interface{} {
+	return newStringBuilder()
+}}
+
+func newStringBuilder() stringBuilder {
+	return &strings.Builder{}
+}
+
+func main() {
+	i := builderPool.Get()
+	sb := i.(stringBuilder)
+	_, _ = sb.WriteString("hello")
+
+	println(sb.String())
+
+	builderPool.Put(i)
+}
+
+// Output:
+// hello

--- a/interp/run.go
+++ b/interp/run.go
@@ -297,6 +297,9 @@ func typeAssert(n *node, withResult, withOk bool) {
 				}
 				return next
 			}
+			if c0.typ.cat == valueT {
+				valf = reflect.ValueOf(v)
+			}
 			if v.node.typ.id() == typID {
 				if withResult {
 					value0(f).Set(valf)

--- a/interp/value.go
+++ b/interp/value.go
@@ -65,15 +65,17 @@ func genValueBinMethodOnInterface(n *node, defaultGen func(*frame) reflect.Value
 			nod = vi.node
 		}
 
-		if nod == nil {
+		if nod == nil || nod.typ.rtype == nil {
 			return defaultGen(f)
 		}
 
-		typ := nod.typ
-		if typ.node != nil || typ.cat != valueT {
+		// Try to get the bin method, if it doesnt exist, fall back to
+		// the default generator function.
+		meth, ok := nod.typ.rtype.MethodByName(c0.child[1].ident)
+		if !ok {
 			return defaultGen(f)
 		}
-		meth, _ := typ.rtype.MethodByName(c0.child[1].ident)
+
 		return meth.Func
 	}
 }

